app/src/main/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java

```diff
@@ -41,33 +43,71 @@
     }
 
     private IAuthenticationManager mUserAccountRepository;
+    private IAsyncExecutor mAsyncExecutor;
+    private IMainExecutor mMainExecutor;
+    private Callback mCallback;
+    private Credentials mCredentials;
 
-    public LoginUseCase(IAuthenticationManager userAccountRepository) {
+    public LoginUseCase(IAuthenticationManager userAccountRepository, IMainExecutor mainExecutor,
+            IAsyncExecutor asyncExecutor) {
         mUserAccountRepository = userAccountRepository;
+        mMainExecutor = mainExecutor;
+        mAsyncExecutor = asyncExecutor;
     }
 
-    public void execute(Credentials credentials, final Callback callback) {
-        
-        mUserAccountRepository.login(credentials,
+    @Override
+    public void run() {
+        mUserAccountRepository.login(mCredentials,
                 new IRepositoryCallback<UserAccount>() {
                     @Override
                     public void onSuccess(UserAccount userAccount) {
-                        callback.onLoginSuccess();
+                        mCallback.onLoginSuccess();
                     }
 
                     @Override
                     public void onError(Throwable throwable) {
                         if (throwable instanceof MalformedURLException
                                 || throwable instanceof UnknownHostException) {
-                            callback.onServerURLNotValid();
+                            notifyOnServerURLNotValid();
                         } else if (throwable instanceof InvalidCredentialsException) {
-                            callback.onInvalidCredentials();
+                           notifyOnInvalidCredentials();
                         } else if (throwable instanceof NetworkException) {
-                            callback.onNetworkError();
+                            notifyOnNetworkError();
                         }
                     }
                 });
     }
 
+    private void notifyOnServerURLNotValid(){
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onServerURLNotValid();
+            }
+        });
+    }
+    private void notifyOnInvalidCredentials(){
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onInvalidCredentials();
+            }
+        });
+    }
+    private void notifyOnNetworkError(){
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onNetworkError();
+            }
+        });
+    }
+
+    public void execute(Credentials credentials, final Callback callback) {
```

Comment:
Please move execute method before run method for we have a code more readable according to execution flow